File: LeetCode/WordLadder.java
package com.solution;

import java.util.*;
/*
127. Word Ladder
https://leetcode.com/problems/word-ladder/description/
 */

public class WordLadder {
    int change = Integer.MAX_VALUE;

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>();
        helper(beginWord, endWord, wordList, 0, -1, set);
        System.out.print((change == Integer.MAX_VALUE) ? 0 : change);
        return (change == Integer.MAX_VALUE) ? 0 : change;

    }

    private boolean helper(String beginWord, String endWord, List<String> wordList, int i, int k, Set<String> set) {
        if (endWord.equals(beginWord)) {
            change = Math.min(i + 1, change);
            return true;
        } else {
            for (int j = 0; j < wordList.size(); j++) {
                if (set.contains(wordList.get(j)) || j == k) continue;

                if (checkChange(beginWord, wordList.get(j))) {

                    if (set.add(wordList.get(j))) {
                        helper(wordList.get(j), endWord, wordList, i + 1, j, set);
                        set.remove(wordList.get(j));
                    }
                }
            }
        }
        return false;
    }

    private boolean checkChange(String s1, String s2) {
        boolean onceMatch = false;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                if (onceMatch) {
                    return false;
                }
                onceMatch = true;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        WordLadder obj = new WordLadder();
        List<String> lst = new ArrayList<>(Arrays.asList("hot", "dog"));
        /*
        "hot"
        "dog"
        ["hot","dog","dot"]
         */

        String begin = "hot";
        String end = "dog";

        List<String> lst1 = new ArrayList<>(Arrays.asList("hot", "dot", "dog", "lot", "log", "cog"));
        String begin1 = "hit";
        String end1 = "cog";

        String begin2 = "hot";
        String end2 = "dog";
        List<String> lst2 = new ArrayList<>(Arrays.asList("hot", "dog", "dot"));

        String begin3 = "a";
        String end3 = "c";
        List<String> lst3 = new ArrayList<>(Arrays.asList("a", "b", "c"));
         obj.ladderLength(begin, end, lst);
//        obj.ladderLength(begin1, end1, lst1);
//        obj.ladderLength(begin2, end2, lst2);
//        obj.ladderLength(begin3, end3, lst3);

    }
}

File: LeetCode/WordLadder.py
'''

Word Ladder

A transformation sequence from word beginWord to word endWord using a dictionary wordList is a sequence of words such that:

The first word in the sequence is beginWord.
The last word in the sequence is endWord.
Only one letter is different between each adjacent pair of words in the sequence.
Every word in the sequence is in wordList.
Given two words, beginWord and endWord, and a dictionary wordList, return the number of words in the shortest transformation sequence from beginWord to endWord, or 0 if no such sequence exists.

 

Example 1:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
Output: 5
Explanation: One shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog" with 5 words.
Example 2:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
Output: 0
Explanation: The endWord "cog" is not in wordList, therefore there is no possible transformation.
 

Constraints:

1 <= beginWord.length <= 10
endWord.length == beginWord.length
1 <= wordList.length <= 5000
wordList[i].length == beginWord.length
beginWord, endWord, and wordList[i] consist of lowercase English letters.
beginWord != endWord
All the strings in wordList are unique.

'''

from collections import deque
from collections import namedtuple

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        
        if endWord not in wordList:
            return 0
        
        q = deque()
        Element = namedtuple('Element',['word','lst','count']) 
        q.append(Element(beginWord, wordList, 1))
        visited = set()
        
        while len(q) > 0:
            element = q.popleft()
            word = element.word
            if word in visited:
                continue
            for possible in element.lst:
                if possible not in visited and self._single_change(word, possible):
                    if possible == endWord:
                        return element.count + 1
                    q.append(Element(possible, [x for x in element.lst if x != possible], element.count + 1))
            visited.add(word)
        
        return 0
    
    def _single_change(self, word1, word2):
        changed = False
        for x, y in zip(word1, word2):
            if changed and x != y:
                return False
            
            if x != y:
                changed = True
        
        return True
        
